shell/platform/fuchsia/dart_runner/dart_runner.h

```diff
@@ -26,7 +26,8 @@ class DartRunner : public fuchsia::sys::Runner {
       ::fidl::InterfaceRequest<fuchsia::sys::ComponentController> controller)
       override;
 
-  std::unique_ptr<sys::ComponentContext> context_;
+  // Not owned by DartRunner.
+  sys::ComponentContext* context_;
```

Comment:
Hrm. The raw pointer is less-than-ideal here. Based on https://github.com/dart-lang/sdk/issues/41523, looks like the source of this is a singleton static that was added here:
https://dart-review.googlesource.com/c/sdk/+/136524

Is there any way we could avoid a process-wide static? This makes testing, dynamic loading/unloading, multi-tenancy trickier. Granted, it sounds like multi-tenancy is unlikely to be an issue if the goal is a process-wide value.

I'd really like to avoid us passing around raw pointers. The fact that it's being injected via a ctor suggests that this is a generic unowned pointer which we'd typically do via a weak pointer to provide a bit of safety in the runner code.

/cc @chinmaygarde for thoughts

Reply:
Hm, not sure why my comment didn't get recorded.

The goal here is a process-wide value, because it *is* a process-wide value.  In Fuchsia, you can not have a process with two of these that are both valid.  When you construct an instance of `sys::ComponentContext`, it consumes a predefined handle used for initializing the context properly, so followup calls to `sys::ComponentContext::Create()` will produce defunct component context objects.

The second issue that I think requires this to be static is that there is no way to pass this platform-specific value into dart's `os_fuchsia.cc`, since there are no slots providing a way to pass this top-level object into os-specific code in dart.  I would love to be wrong here, but I couldn't see a way around it.

Reply:
> When you construct an instance of sys::ComponentContext, it consumes a predefined handle used for initializing the context properly, so followup calls to sys::ComponentContext::Create() will produce defunct component context objects.

Then this seem like something neither the runner nor the runtime should be responsible for creating and owning the pointer to the singleton to. For example, what happens in a process that may or may not need to the Dart runtime to perform its duties? Someone else will have to create the singleton but once the Dart VM comes along, it will expect to be successfully able to create it for its own "ownership". If this object it truly a singleton, can it be put in some spot in the SDK that everyone can refer to instead of now implicitly making sure a runtime compnent owns it?

Reply:
What I can do at the moment is to remove the need for the runner to own it here, but instead simply request it from `dart::ComponentContext()` at the point it is needed.  

I also think it worthwhile for the SDK to change, but that's going to take a while, as it could be a breaking change.

How about I remove this pointer altogether and call `dart::ComponentContext()` directly where needed?
Once (or if, since I don't own that decision) a singleton is available in the SDK, we can change the call sites to invoke the SDK function.

Reply:
> or if, since I don't own that decision

Who would be the owner of that decision? Can we file an issue to track that down.

Once this does get into the SDK, existing uses of `dart::ComponentContext()` will be breaking. Not sure when that was added but it seems like an odd place to initialize a singleton.

Reply:
I'm guessing the component framework team in Fuchsia would own it.   I added `dart::ComponentContext()`.  To explain what I'm trying to do: I want to improve i18n test coverage for dart and flutter in Fuchsia, and part of that is to add inspect metrics into the dart VM, for white-box testing.  Since I want to observe the state of the `OS` object, there is no way for me to obtain the currently valid `ComponentContext`, and hence I need to get it out of band.

Once a SDK version of this is available we can shunt `dart::ComponentContext()` to that function, then remove all uses.

Reply:
> Not sure when that was added but it seems like an odd place to initialize a singleton.

Why is this an odd place to initialize a singleton?  It is a fuchsia-platform specific module.

Reply:
> Once a SDK version of this is available we can shunt dart::ComponentContext() to that function, then remove all uses.

SGTM.

> Why is this an odd place to initialize a singleton? It is a fuchsia-platform specific module.

Now, the runtime is responsible for creating and vending the pointer to the singleton. Components that don't care about the runtime but need to access this same fuchsia platform specific functionality now have to ask the runtime for the same. And if the some other component happens to initialize it, the runtime is out of luck. I suppose this is all just the usual fallout of using singletons. But, if we are tracking and following up on the SDK improvements, this is fine for now.

Reply:
> > Once a SDK version of this is available we can shunt dart::ComponentContext() to that function, then remove all uses.
> SGTM.

Thanks!

> > Why is this an odd place to initialize a singleton? It is a fuchsia-platform specific module.
> Now, the runtime is responsible for creating and vending the pointer to the singleton. Components that don't care about the runtime but need to access this same fuchsia platform specific functionality now have to ask the runtime for the same. And if the some other component happens to initialize it, the runtime is out of luck. I suppose this is all just the usual fallout of using singletons. But, if we are tracking and following up on the SDK improvements, this is fine for now.

Noted.  What I think makes this situation a bit less onerous than the general case, is that:

1. the "if some other component happens to initialize it first we're out of luck" is already the case today. It is probably only ever correct to request the component context in your `main` function and pass the context down to where it is needed.
2. all platform-specific things for Fuchsia are sequestered where they are easy to find. So once SDK solution is in place we can make the change.  It has to be done carefully, but it seems that any transition between here and there won't make things worse than they are today.